README.md
[![Circle CI](https://circleci.com/gh/rackspace-cookbooks/rackspace_nginx_php.svg?style=svg)](https://circleci.com/gh/rackspace-cookbooks/rackspace_nginx_php)

# rackspace_nginx_php-cookbook

A cookbook to provide a web server able to serve php pages with Nginx and PHP fpm.
It relies on [nginx cookbook](https://github.com/miketheman/nginx) and [php-fpm](https://github.com/yevgenko/cookbook-php-fpm). Those cookbooks are pinned on well known working minor version to prevent breaking changes.
In addition (even if this is not a requirement), the cookbook will install php packages through [PHP cookbook](https://github.com/opscode-cookbooks/php). Indeed most of the time you will need `php::default` in your role which will conflict with the `php-fpm` package if they are different.
You can disable the installation of php packages with `node['rackspace_nginx_php']['php_packages_install']['enable']`.


## *** 
## NOTE: Support for PHP 5.4 was dropped in v1.0.0
## ***

## Supported Platforms

* Centos 6.7
* Ubuntu 12.04
* Ubuntu 14.04

## Supported PHP versions

* 5.5
* 5.6

## Attributes

* `node['rackspace_nginx_php']['php_version']` : Which PHP version to install, default to PHP 5.6
* `node['rackspace_nginx_php']['php-fpm']['default_pool']['enable']` : Should it enable a default PHP-FPM pool which listens on a unix socket
* `node['rackspace_nginx_php']['nginx']['default_site']['enable']` : Should it enable a default Nginx site which passes requests for php files to the default php-fpm pool
* `node['rackspace_nginx_php']['nginx']['default_site']['cookbook']` : Where to find the template for the default Nginx site configuration , default to `rackspace_nginx_php cookbook`
* `node['rackspace_nginx_php']['nginx']['default_site']['template']` : The name of the template for the default Nginx site configuration , default to `default.conf.erb`

## Usage

Place a dependency on the rackspace_nginx_php cookbook in your cookbook's metadata.rb
```
depends 'rackspace_nginx_php'
```
Then, add `rackspace_nginx_php::default` to your runlist

```
# myrecipe.rb
include_recipe 'rackspace_nginx_php::default'
```

or

```
# roles/myrole.rb
name "myrole"
description "nginx and php role"
run_list(
  "rackspace_nginx_php::default"
)
```

You can change any of the `nginx`,`php-fpm` and `php` cookbook attributes to tune rackspace_nginx_php configuration.
** However you should not change ** `['php-fpm']['package_name']`,`['php-fpm']['service_name']` or `['php']['packages']` (as they are part of this cookbook logic) without checking it works.

## In scope

The goal of this library is to do the basic configuration to serve PHP pages through Nginx. It will only configure `Nginx` and a default site and pool, users are free to use that or disable it and create their own Nginx sites and PHP-FPM pools.

More in details it :

* Installs and configure Nginx web server
* Installs and configure php-fpm
* Installs and configure php
* Configures Nginx to serve php pages through php-fpm
* Gets the correct packages and change the configuration according to the php version

## Out of scope

Virtual Host are not managed by this cookbook, the configuration provided by this cookbook should not prevent users to extend php or php-fpm configuration.
As many features as possible should have a flag to enable/disable them, it will allow to enjoy some parts of the work done by this cookbook (get the correct packages by example) but still be able to configure your own php-fpm pools.


### Examples
#### Nginx and PHP 5.5

```
node.default['rackspace_nginx_php']['php_version'] = '5.5'
include_recipe 'rackspace_nginx_php::default'
```

#### Nginx and PHP 5.6

```
include_recipe 'rackspace_nginx_php::default'
```

#### Nginx and PHP 5.6 without the default Nginx site and PHP-FPM pool - Using attribute from upstream to define the pool

You will have to add your own Nginx site and PHP-FPM pool, here is an example using the upstream PHP-FPM attributes for the pool and a custom template for the Nginx site.

```
node.default['rackspace_nginx_php']['nginx']['default_site']['enable'] = false
node.default['rackspace_nginx_php']['php-fpm']['default_pool']['enable'] = false

# Create a php-fpm pool through the attribute exposed from upstream
node.default['php-fpm']['pools'] = {
  override: {
    enable: 'true',
    process_manager: 'dynamic',
    max_requests: 5000
  }
}

include_recipe 'rackspace_nginx_php::default'

# Create a nginx site to go with it (the template must be created by the user)
template "#{node['nginx']['dir']}/sites-available/override.conf" do
  source 'override.conf.erb'
  owner 'root'
  group 'root'
  mode '0644'
  action :create
end

nginx_site 'override.conf' do
  enable true
end

```

#### Nginx and PHP 5.6 without the default Nginx site and PHP-FPM pool - Using the php_fpm_pool definition to add a pool

You will have to add your own Nginx site and PHP-FPM pool, here is an example using the upstream PHP-FPM definition for the pool and a custom template for the Nginx site.

```
node.default['rackspace_nginx_php']['nginx']['default_site']['enable'] = false
node.default['rackspace_nginx_php']['php-fpm']['default_pool']['enable'] = false

# Create a php-fpm pool with default settings using the php_fpm_pool definition
php_fpm_pool 'override'

include_recipe 'rackspace_nginx_php::default'

# Create a nginx site to go with it (the template must be created by the user)
template "#{node['nginx']['dir']}/sites-available/override.conf" do
  source 'override.conf.erb'
  owner 'root'
  group 'root'
  mode '0644'
  action :create
end

nginx_site 'override.conf' do
  enable true
end

```


## References

* [Nginx cookbook](https://github.com/miketheman/nginx)
* [PHP-fpm cookbook](https://github.com/yevgenko/cookbook-php-fpm)
* [PHP cookbook](https://github.com/opscode-cookbooks/php)


## Contributing

1. Fork the repository on Github
2. Create a named feature branch (i.e. `add-new-recipe`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request

## License and Authors

Authors:: Kostas Georgakopoulos (kostas.georgakopoulos@rackspace.co.uk), Julien Berard (julien.berard@rackspace.co.uk)


CHANGELOG.md
rackspace_nginx_php CHANGELOG
==================

1.0.0
-----
- Release as v1.0.0
- Dropped support for PHP 5.4
- Updated unit tests

0.0.5
-----
- Support for CentOS 6.7 instead of 6.6

0.0.4
-----
- Fix for rubocop tests

0.0.3
-----
- Improvements on integration tests

0.0.2
-----
- Pinned upstream cookbook versions
- Added 'supports' statements to metadata
- Various improvements and clean up on tests

0.0.1
-----
- First version


test/unit/spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'
require_relative 'rackspace_nginx_php_shared'

::LOG_LEVEL = ENV['CHEFSPEC_LOG_LEVEL'] ? ENV['CHEFSPEC_LOG_LEVEL'].to_sym : :fatal

def stub_resources
  stub_command('which nginx').and_return(true)
  stub_command('test -d /etc/php-fpm.d || mkdir -p /etc/php-fpm.d').and_return(true)
  stub_command('test -d /etc/php5/fpm/pool.d || mkdir -p /etc/php5/fpm/pool.d').and_return(true)
end

def node_resources(_node)
end

at_exit { ChefSpec::Coverage.report! }


test/unit/spec/ubuntu1404_options.rb
  UBUNTU1404_SERVICE_OPTS = {
    log_level: LOG_LEVEL,
    platform: 'ubuntu',
    version: '14.04'
  }


test/unit/spec/rackspace_nginx_php_shared.rb
shared_examples_for 'Nginx' do |enabled_suite, disabled_suite|
  it 'includes nginx required recipes' do
    expect(chef_run).to include_recipe('nginx::default')
  end
  it 'creates the enabled site document root' do
    expect(chef_run).to create_directory("/var/www/#{enabled_suite}").with(
      owner: 'root',
      group: 'root',
      mode: '0755',
      recursive: true
    )
  end
  it 'does not create the disabled site document root' do
    expect(chef_run).to_not create_directory("/var/www/#{disabled_suite}")
  end
  it 'creates the enabled site configuration' do
    [
      'server {',
      'listen        80;',
      'server_name   Fauxhai;',
      "access_log    /var/log/nginx/#{enabled_suite}.access.log;",
      "error_log     /var/log/nginx/#{enabled_suite}.error.log;",
      'location / {',
      'index   index.html index.htm;',
      'location ~* \.php$ {',
      'try_files       $uri = 404;',
      'include         fastcgi_params;',
      "fastcgi_pass    unix:/var/run/php-fpm-#{enabled_suite}.sock;",
      'fastcgi_index   index.php;',
      'fastcgi_param   SCRIPT_FILENAME  $document_root$fastcgi_script_name;'
    ].each do |line|
      expect(chef_run).to render_file("/etc/nginx/sites-available/#{enabled_suite}.conf").with_content(line)
    end
  end
  it 'does not create the disabled site configuration' do
    expect(chef_run).to_not render_file("/etc/nginx/sites-available/#{disabled_suite}.conf")
  end
  it 'enables the site' do
    expect(chef_run).to enable_nginx_site("#{enabled_suite}")
  end
end

shared_examples_for 'PHP-FPM' do |platform, suite|
  it 'includes php-fpm default recipe' do
    expect(chef_run).to include_recipe('php-fpm::default')
  end
  it 'creates the default php-fpm pool' do
    # We cannot test the php_fpm_pool as it is a definition not a resource
    if platform == 'redhat'
      expect(chef_run).to render_file("/etc/php-fpm.d/#{suite}.conf")
    elsif platform == 'ubuntu'
      expect(chef_run).to render_file("/etc/php5/fpm/pool.d/#{suite}.conf")
    else
      expect(chef_run).to render_file("/etc/php5/fpm/pool.d/#{suite}.conf")
    end
  end
end

shared_examples_for 'Yum IUS repo' do
  it 'configures the appropriate repository' do
    expect(chef_run).to include_recipe('yum-ius')
  end
end

shared_examples_for 'APT php repo' do |version|
  it 'configures the appropriate repository' do
    expect(chef_run).to add_apt_repository("php-#{version}")
  end
end

shared_examples_for 'PHP and PHP-fpm packages version 5.5 CENTOS' do
  it 'installs the correct php and php-fpm packages' do
    ['php55u-devel', 'php55u-cli', 'php55u-pear', 'php55u-fpm'].each do |pkg|
      expect(chef_run).to install_package(pkg)
    end
  end
end

shared_examples_for 'PHP and PHP-fpm packages version 5.6 CENTOS' do
  it 'installs the correct php and php-fpm packages' do
    ['php56u-devel', 'php56u-cli', 'php56u-pear', 'php56u-fpm'].each do |pkg|
      expect(chef_run).to install_package(pkg)
    end
  end
end

shared_examples_for 'PHP-fpm packages without PHP packages, version 5.6 CENTOS' do
  it 'installs the correct php and php-fpm packages' do
    expect(chef_run).to install_package('php56u-fpm')
    expect(chef_run).to_not install_package('php56u-pear')
  end
end

shared_examples_for 'PHP and PHP-fpm packages version 5.5 UBUNTU' do
  it 'installs the correct php and php-fpm packages' do
    %w(php5-cgi php5 php5-dev php5-cli php-pear php5-fpm).each do |pkg|
      expect(chef_run).to install_package(pkg)
    end
  end
end

shared_examples_for 'PHP and PHP-fpm packages version 5.6 UBUNTU' do
  it 'installs the correct php and php-fpm packages' do
    %w(php5-cgi php5 php5-dev php5-cli php-pear php5-fpm).each do |pkg|
      expect(chef_run).to install_package(pkg)
    end
  end
end

shared_examples_for 'PHP-fpm packages without PHP packages, version 5.6 UBUNTU' do
  it 'installs the correct php and php-fpm packages' do
    expect(chef_run).to install_package('php5-fpm')
    expect(chef_run).to_not install_package('php5-cgi')
  end
end


recipes/default.rb
# Encoding: utf-8
#
# Cookbook Name:: rackspace_nginx_php
# Recipe:: default
#
# Copyright 2014, Rackspace
#

# We only support 5.5 and 5.6
if node['rackspace_nginx_php']['php_version'] != '5.5' && node['rackspace_nginx_php']['php_version'] != '5.6'
  Chef::Application.fatal!("PHP version #{node['rackspace_nginx_php']['php_version']} is not supported")
end

include_recipe 'chef-sugar'

# NGINX
# Disable the default site from the upstream nginx cookbook.
# We do that because the template source is hardcoded in upstream and we cannot override it.
# We will create owr own.
node.default['nginx']['default_site_enabled'] = false

include_recipe 'nginx'

# Create the document root to use for the default site we create
directory '/var/www/default' do
  owner 'root'
  group 'root'
  mode '0755'
  recursive true
  action :create
  only_if { node['rackspace_nginx_php']['nginx']['default_site']['enable'] }
end

# Configuration for our default site
template "#{node['nginx']['dir']}/sites-available/default.conf" do
  source node['rackspace_nginx_php']['nginx']['default_site']['template']
  cookbook node['rackspace_nginx_php']['nginx']['default_site']['cookbook']
  owner 'root'
  group 'root'
  mode '0644'
  action :create
  only_if { node['rackspace_nginx_php']['nginx']['default_site']['enable'] }
end

# Enable (or not) our default site
nginx_site 'default.conf' do
  enable node['rackspace_nginx_php']['nginx']['default_site']['enable']
end

# PHP-FPM
# repo dependencies for php-fpm
#
php_fpm = {
  'rhel' => {
    '5.5' => {
      'php_packages' => ['php55u-pear', 'php55u-devel', 'php55u-cli', 'php55u-pear'],
      'php_fpm_package' => 'php55u-fpm',
      'service' => 'php-fpm'
    },
    '5.6' => {
      'php_packages' => ['php56u-pear', 'php56u-devel', 'php56u-cli', 'php56u-pear'],
      'php_fpm_package' => 'php56u-fpm',
      'service' => 'php-fpm'
    }
  },
  'debian' => {
    '5.5' => {
      'repo'    => 'http://ppa.launchpad.net/ondrej/php5/ubuntu',
      'php_packages' => ['php5-cgi', 'php5', 'php5-dev', 'php5-cli', 'php-pear'],
      'php_fpm_package' => 'php5-fpm',
      'service' => 'php5-fpm'
    },
    '5.6' => {
      'repo'    => 'http://ppa.launchpad.net/ondrej/php5-5.6/ubuntu',
      'php_packages' => ['php5-cgi', 'php5', 'php5-dev', 'php5-cli', 'php-pear'],
      'php_fpm_package' => 'php5-fpm',
      'service' => 'php5-fpm'
    }
  }
}

# Platform PHP repositories
# RHEL
if platform_family?('rhel')
  include_recipe 'yum-ius'
elsif platform_family?('debian')
  # DEBIAN
  include_recipe 'apt'
  # using http://ppa rather than ppa: to be sure it passes firewall
  apt_repository "php-#{node['rackspace_nginx_php']['php_version']}" do
    uri          php_fpm[node['platform_family']][node['rackspace_nginx_php']['php_version']]['repo']
    keyserver    'hkp://keyserver.ubuntu.com:80'
    key          '14AA40EC0831756756D7F66C4F4EA0AAE5267A6C'
    components   ['main']
    distribution node['lsb']['codename']
  end
end

# PHP-FPM
# Set the correct php-fpm packages to install
node.default['php-fpm']['package_name'] = php_fpm[node['platform_family']][node['rackspace_nginx_php']['php_version']]['php_fpm_package']
node.default['php-fpm']['service_name'] = php_fpm[node['platform_family']][node['rackspace_nginx_php']['php_version']]['service']

# Set the correct user for php-fpm
node.default['php-fpm']['user'] = node['nginx']['user']
node.default['php-fpm']['group'] = node['nginx']['group']

include_recipe 'php-fpm::default'

# Create (or not) our default pool
php_fpm_pool 'default' do
  enable node['rackspace_nginx_php']['php-fpm']['default_pool']['enable']
end

# PHP
# Set the correct php packages to install
if node['rackspace_nginx_php']['php_packages_install']['enable']
  node.default['php']['packages'] = php_fpm[node['platform_family']][node['rackspace_nginx_php']['php_version']]['php_packages']
  include_recipe 'php::default'
  # Emptying php packages, to be sure if another cookbook tries to include php::default it
  # will not install unwanted php version. If a user decided to use override we assume he knows what he is doing.
  node.default['php']['packages'] = []
end

# TODO
# * downgrade / upgrade Nginx if required
# Refactor mapping outside of the recipe
